core/src/main/scala/org/apache/spark/shuffle/hash/HashShuffleManager.scala

```diff
@@ -50,15 +53,30 @@ private[spark] class HashShuffleManager(conf: SparkConf) extends ShuffleManager
   }
 
   /** Get a writer for a given partition. Called on executors by map tasks. */
-  override def getWriter[K, V](handle: ShuffleHandle, mapId: Int, context: TaskContext)
-      : ShuffleWriter[K, V] = {
-    new HashShuffleWriter(
-      shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]], mapId, context)
+  override def getWriter[K, V](
+      handle: ShuffleHandle,
+      mapId: Int,
+      stageAttemptId: Int,
+      context: TaskContext): ShuffleWriter[K, V] = {
+    addShuffleAttempt(handle.shuffleId, stageAttemptId)
+    new HashShuffleWriter(shuffleBlockResolver, handle.asInstanceOf[BaseShuffleHandle[K, V, _]],
```

Comment:
each ShuffleManager must call `addShuffleAttempt` in `getWriter`, to ensure that all attempts get cleaned up when the shuffle is unregistered.

now that I'm writing this I'm wondering ... should I make a new abstract method `getWriterHelper` or something, which only has to get the writer, and change `ShuffleManger.getWriter` to be a concrete implementation like:

```
def getWriter[K, V](
     handle: ShuffleHandle,
     mapId: Int,
     stageAttemptId: Int,
     context: TaskContext): ShuffleWriter[K, V] = {
  addShuffleAttempt(handle.shuffleId, stageAttemptId)
  getWriterHelper(handle, mapId, stageAttemptId, context)
}
```
?

though I'm not so sure it will really help b/c of the way `UnsafeShuffleManager` delegates to `SortShuffleManager`, and each implementation has to implement other cleanup logic in any case ...